proposals/0186-transaction-data-size-specification.md

```diff
@@ -0,0 +1,143 @@
+programs provided as instruction accounts.
+* Transaction accounts list: all accounts for the transaction, which includes
+top-level programs, the fee-payer, all instruction accounts, and any extra
+accounts added to the list but not used for any purpose.
+
+## Detailed Design
+
+The proposed algorithm is as follows:
+
+1. Every account explicitly included on the transaction accounts list is counted
+once and only once.
+2. A valid program owned by LoaderV3 also includes the size of its programdata.
+3. Other than point 2, no accounts are implicitly added to the total data size.
+
+Transactions may include a
+`ComputeBudgetInstruction::SetLoadedAccountsDataSizeLimit` instruction to define
+a data size limit for the transaction. Otherwise, the default limit is 64MiB
+(`64 * 1024 * 1024` bytes).
+
+If a transaction exceeds its data size limit, account loading is aborted and the
+transaction is failed. Fees will be charged once
+`enable_transaction_loading_failure_fees` is enabled.
+
+Adding required loaders to transaction data size is abolished. They are treated
+the same as any other account: counted if on the transaction accounts list, not
+counted otherwise.
+
+Read-only and writable accounts are treated the same. In the future, when direct
+mapping is enabled, this SIMD may be amended to count them differently.
+
+As a consequence of 1 and 2, for LoaderV3 programs, programdata is counted twice
+if a transaction includes both programdata and the program account itself in the
+accounts list, unless the program is not valid for execution. This is partly
+done for ease of implementation: we always want to count programdata when the
+program is included, and there is no reason for any transaction to include both
+accounts except during initial deployment, in which case the program is not yet
+valid.
+
+We include programdata size in program size for LoaderV3 programs because in
+nearly all cases a transaction will include the program account (the only way to
+invoke the program) and will not include the programdata account because
+including it serves no purpose. Including the program account forces an
+unconditional load of the programdata account because it is required to compile
+the program for execution. Therefore we always count it, even when the program
+is an instruction account.
+
+There is no special handling for programs owned by the native loader, LoaderV1,
+or LoaderV2.
+
+Account size for programs owned by LoaderV4 is left undefined. This SIMD should
+be amended to define the required semantics before LoaderV4 is enabled on any
+network.
+
+## Alternatives Considered
+
+* Transaction data size accounting is already enabled, so the null option is to
+enshrine the current Agave behavior in the protocol. This is undesirable because
+the current behavior is highly idiosyncratic, and LoaderV3 program sizes are
+routinely undercounted.
+* Builtin programs are backed by accounts that only contain the program name as
```

Comment:
it's also in the works to make (some/most) builtins just normal programs, so adding any special cases here would probably be nullified in the future anyway